qendpoint-store/src/main/java/com/the_qa_company/qendpoint/utils/VariableToIdSubstitution.java

```diff
@@ -41,6 +44,50 @@ public void optimize(TupleExpr tupleExpr, Dataset dataset, BindingSet bindings)
 
 	protected class Substituor extends AbstractQueryModelVisitor<RuntimeException> {
 
+		@Override
+		public void meet(StatementPattern node) throws RuntimeException {
+			Var subjectVar = node.getSubjectVar();
+			if (subjectVar != null && subjectVar.isAnonymous() && subjectVar.hasValue()) {
+				long id = converter.subjectToID(((Resource) subjectVar.getValue()));
+				if (id != -1) {
+					Var var1 = new Var(subjectVar.getName(), converter.idToSubjectHDTResource(id), true,
+							subjectVar.isConstant());
+					node.replaceChildNode(subjectVar, var1);
+				}
+			}
+
+			Var predicateVar = node.getPredicateVar();
+			if (predicateVar != null && predicateVar.isAnonymous() && predicateVar.hasValue()) {
+				long id = converter.predicateToID(((IRI) predicateVar.getValue()));
+				if (id != -1) {
+					Var var1 = new Var(predicateVar.getName(), converter.idToPredicateHDTResource(id), true,
+							predicateVar.isConstant());
+					node.replaceChildNode(predicateVar, var1);
+				}
+			}
+
+			Var objectVar = node.getObjectVar();
+			if (objectVar != null && objectVar.isAnonymous() && objectVar.hasValue()) {
+				long id = converter.objectToID((objectVar.getValue()));
+				if (id != -1) {
+					Var var1 = new Var(objectVar.getName(), converter.idToObjectHDTResource(id), true,
+							objectVar.isConstant());
+					node.replaceChildNode(objectVar, var1);
+				}
+			}
+
+			Var contextVar = node.getContextVar();
+			if (contextVar != null && contextVar.isAnonymous() && contextVar.hasValue()) {
+				long id = converter.contextToID((((Resource) contextVar.getValue())));
+				if (id != -1) {
+					Var var1 = new Var(contextVar.getName(), converter.idToGraphHDTResource(id), true,
+							contextVar.isConstant());
+					node.replaceChildNode(contextVar, var1);
+				}
+			}
+
+		}
+
```

Comment:
Here is the addition where I optimize each var in the StatementPattern. For some reason the code in `meet(Var var)` didn't manage to make the predicate have the correct position. When working on the StatementPattern we already know which position every var has.

Reply:
Did you try to use a constant outside a triple to see if it works? For example in a filter.

Reply:
I haven't tried that. But I assume it would match with the existing meet method for vars. I haven't removed that one, since the meet on the statement pattern doesn't work for things like BIND or FILTER.